non-global Rollbar clients

fabiokung opened the pull request:
Also introduce the `Client` interface, so code using this library can be tested (i.e.: others can mock the rollbar client).

These changes are (mostly) backwards compatible. I kept the global functions (e.g.: `rollbar.Message(...)`), backed by a singleton global `Client` (`rollbar.std`), similarly to what the `log` package does with `log.Logger`.

@tysontate would you be interested in pulling this back upstream?

fabiokung committed:
add rollbar.New(), allowing for multiple clients and non-global use

Also introduce the Client interface so code using this library
can be more testable (i.e.: others can mock the rollbar client).

fabiokung commented:
/cc @ryanbrainard

ryanbrainard commented on rollbar.go:
```diff
@@ -31,53 +31,101 @@ const (
+}
+
+func New(token, environment, codeVersion, serverHost, serverRoot string) Client {
+	return newRollbar(token, environment, codeVersion, serverHost, serverRoot)
```
What's the advantage of having both `New()` and `newRollbar()`? Can this be inlined?

fabiokung replied:
`New()` exposes the public interface only. `newRollbar()` is just a helper to build the concrete struct. Note the return types.

ryanbrainard commented on rollbar.go:
```diff
@@ -31,53 +31,101 @@ const (
+}
+
+var (
+	hostname, _ = os.Hostname()
```
any reason `os.Hostname()` isn't part of `newRollbar()`?

fabiokung replied:
I just felt that querying `os.Hostname()` every time a Client is built was necessary. Thoughts?

ryanbrainard commented:
What do you think about breaking out all of the `Std` wrappers into a separate file? I think that might make it clearer what's going on.

ryanbrainard commented:
Remember to bump the client version

tysonmote commented:
@fabiokung I'd definitely be interested in pulling this upstream.

This package probably needs some other cleanup. It was originally written for a single long-running app that used a single Rollbar project, but it's clear that it's being used in slightly different ways that I could do a better job of supporting.

I'll keep an eye on this thread and then merge upstream whatever makes sense. Hopefully everything. :)

tysonmote commented:
Also — I think I want to make all my API-breaking changes at once. What do you think about removing `Buffer` and the async-by-default behavior? I think it makes more sense to be synchronous by default. This has a few advantages:
- Users can choose whether it makes sense to send messages asynchronously or not. All they need to do is wrap their call in a goroutine.
- It removes some needless complexity from the library.
- Users can handle failed Rollbar message sends in a way that makes sense for them. (Retry, escalate, whatever.)

fabiokung commented:
> What do you think about breaking out all of the Std wrappers into a separate file? I think that might make it clearer what's going on.

@ryanbrainard +1, I will push changes soon.

fabiokung commented:
> Remember to bump the client version

Should this be in the PR as well? Or is it better to do after it's merged?

fabiokung commented:
> What do you think about removing Buffer and the async-by-default behavior? I think it makes more sense to be synchronous by default.

@tysontate it seems to me that async by default is a reasonable choice, given that in general you don't want a reporting library like rollbar blocking the critical path of any app using it. But I agree there should be a synchronous mode for people that want more control over it.

How about two different implementations of the `Client` interface? One sync, one async, with the default (the `rollbar.Std` singleton) being async?

fabiokung committed:
move the client definition and default impl to a new file

fabiokung committed:
make everything necessary basic of the public interface

and inline newRollbar(), no need to expose the concrete impl anymore

fabiokung commented:
@ryanbrainard how does this look now?

fabiokung committed:
move docs to the public interface

and fix some method names that were wrong in docs

fabiokung committed:
keep the global singleton private

Now that the public interface has everything that's required to be public,
no need to expose this anymore.

ryanbrainard commented on client.go:
```diff
@@ -0,0 +1,323 @@
+
+// filterParams filters sensitive information like passwords from being sent to
+// Rollbar.
+func filterParams(pattern *regexp.Regexp, values map[string][]string) map[string][]string {
```
should `filterParams()` and `flattenValues()` be in `rollbar.go` since they don't have `(c *Rollbar)`?

fabiokung replied:
maybe, but they are currently only being used in `client.go`, so it seemed to make more sense to have them there. They are private anyway.

I'm fine with either. Is that a make-or-break?

ryanbrainard commented:
I find the naming of `Rollbar` and `Client` a little confusing because it's not obvious which one is the interface. The fact that the methods have `(c *Rollbar)` with the variable named `c` but using the `*Rollbar` struct pointer and `Rollbar` being in `client.go` adds to the confusion. I'm not really sure what we should call them, but the way it is now I think it's a little confusing. Otherwise, :+1:

ryanbrainard commented:
@tysontate I agree with @fabiokung that async is a better default. Also, there are a few [more changes](https://github.com/heroku/rollbar/compare/1ecf0ec121f7dd170f951fe8f62f85823c86f82c...197e8a4fd165c96afe325b2297cf3bcaa5b9a6b4) we made in this fork before this particular pull request. If you pull this in, you might want to look at those too. They include filtering out auth headers, extra metadata support, and trace_chain support

fabiokung commented:
I'm always open to suggestions on naming. :-)

I can see your point about it being a bit confusing, but I don't have a better suggestion. Should we go ahead and merge, or wait a bit to see if someone comes up with better names?

fabiokung commented:
> The fact that the methods have (c *Rollbar) with the variable named c but using the *Rollbar struct pointer and Rollbar being in client.go adds to the confusion.

One way to look at it is that `Rollbar` **is a** `Client` (the default implementation), so it makes sense to call it `c` and for it to be in `client.go`.

fabiokung committed:
make explicit that the default implementation is async

the Client interface doesn't require implementations to be async though

fabiokung commented:
@ryanbrainard how about now? (`Rollbar -> AsyncClient`)

fabiokung committed:
redundant comments

ryanbrainard commented:
I like the async impl. Fixes two birds w/ one stone. :+1:

fabiokung commented:
alright, merging in 3, 2, 1...

apg commented:
The original version has the ability to `Wait()` for things to finish. I see a private method `wait()` here, in client.go, but of course I can't call it. I assume this is a typo / oversight?

fabiokung commented:
@apg not a typo. I (somewhat) conciously broke compatibility there, sorry! The new client interface allows for both sync and async implementations and `Wait()` on a sync client wouldn't make much sense.

I can push a patch to make `wait()` public on the concrete implementation, but I don't think it would make much sense to have that as part of the public interface. Would you be OK with explicitly using the concrete type (async implementation) in your case?

apg commented:
Sure. We can do that. I'll remove my branch for fixing now that I know
that's the case. :)

On Thu, Jan 15, 2015 at 3:46 PM, Fabio Kung notifications@github.com
wrote:

> @apg https://github.com/apg not a typo. I (somewhat) conciously broke
> compatibility there, sorry! The new client interface allows for both sync
> and async implementations and Wait() on a sync client wouldn't make much
> sense.
> 
> I can push a patch to make wait() public on the concrete implementation,
> but I don't think it would make much sense to have that as part of the
> public interface. Would you be OK with explicitly using the concrete type
> (async implementation) in your case?
> 
> —
> Reply to this email directly or view it on GitHub
> https://github.com/heroku/rollbar/pull/1#issuecomment-70158455.

apg commented:
@fabiokung I don't see a sync implementation in the current master, which
would actually be the preferred client for my particular use case. The
Async client will come in handy for the next project we're adding rollbar
to.

On Thu, Jan 15, 2015 at 3:56 PM, Andrew Gwozdziewycz apg@heroku.com wrote:

> Sure. We can do that. I'll remove my branch for fixing now that I know
> that's the case. :)
> 
> On Thu, Jan 15, 2015 at 3:46 PM, Fabio Kung notifications@github.com
> wrote:
> 
> > @apg https://github.com/apg not a typo. I (somewhat) conciously broke
> > compatibility there, sorry! The new client interface allows for both sync
> > and async implementations and Wait() on a sync client wouldn't make much
> > sense.
> > 
> > I can push a patch to make wait() public on the concrete implementation,
> > but I don't think it would make much sense to have that as part of the
> > public interface. Would you be OK with explicitly using the concrete type
> > (async implementation) in your case?
> > 
> > —
> > Reply to this email directly or view it on GitHub
> > https://github.com/heroku/rollbar/pull/1#issuecomment-70158455.

fabiokung commented:
@apg correct! Do you want to cotribute a sync implementation of the `Client` interface? :-)

apg commented:
Not sure I have any time to do so, but perhaps.

On Thu, Jan 15, 2015 at 8:09 PM, Fabio Kung notifications@github.com
wrote:

> @apg https://github.com/apg correct! Do you want to cotribute a sync
> implementation of the Client interface? :-)
> 
> —
> Reply to this email directly or view it on GitHub
> https://github.com/heroku/rollbar/pull/1#issuecomment-70193038.
